### ISSUE TITLE
High memory usage on Apple M1 hardware

### ISSUE THREAD

**Issue by hpedrorodrigues** [hadolint/hadolint#729]
- [x] This is a bug report
- [ ] This is a feature request
- [x] I searched existing issues before opening this one

First of all, thanks for this project. 🚀 

### Expected behavior

Hadolint should exit with code 0 when running it against a valid Dockerfile or when running it with the `--version` flag.

### Actual behavior

When I run the following command on a Macbook with the M1 chip

```bash
docker run --rm hadolint/hadolint hadolint --version
```

The container is killed (OOM) and exits with status code 137.

### Steps to reproduce the behavior

Run the following command on a Macbook with the M1 chip

```bash
docker run --rm hadolint/hadolint hadolint --version; echo $?
```

e.g.
![hadolint](https://user-images.githubusercontent.com/10264635/140468500-39044d5a-27d6-413a-8027-90d3d2c39f60.gif)

**Additional environment details (OS, stack version, etc.)**
MacOS Big Sur
Chip: Apple M1

<details>
<summary>docker info</summary>

```text
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Build with BuildKit (Docker Inc., v0.6.3)
  compose: Docker Compose (Docker Inc., v2.0.0)
  scan: Docker Scan (Docker Inc., v0.8.0)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 7
 Server Version: 20.10.8
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runtime.v1.linux runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: e25210fe30a0a703442421b0f60afac609f950a3
 runc version: v1.0.1-0-g4144b63
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.10.47-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 3
 Total Memory: 3.841GiB
 Name: docker-desktop
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Live Restore Enabled: false
```
</details>

Idk if I'm missing something here so please let me know.

**Comment by m-ildefons** [comment]
Hi @hpedrorodrigues ,
I don't have a mac and can't try it out myself, but maybe you can help me debug this a bit more. Hadolint is currently only built for x86_64 platforms. Do other Docker images run ok, especially ones that are only available for x86_64? Have you tried running the [MacOS binary](https://github.com/hadolint/hadolint/releases/download/v2.8.0/hadolint-Darwin-x86_64)?
From your screencap it looks to me a bit as if the container runs against its memory limit. Can you try and run it with a slightly higher limit (e.g. 6Gb)?
The release binaries and docker images are compressed binaries iirc. There has been a bug in the past with the compression software where it would crash with error code 137. If it is not too much, can you try and compile from source, see if that runs?
There are also debian and alpine based docker containers for Hadolint. Maybe one of those runs? `docker run --rm hadolint/hadolint:v2.8.0-alpine hadolint --version`
At any rate, thanks for the bug report.

**Comment by m-ildefons** [comment]
Hi @hpedrorodrigues ,
thanks for spending some time with this. According to the [docker documentation](https://docs.docker.com/desktop/mac/apple-silicon/) running x86_64 images on an M1 mac will utilize an emulation with qemu. I suspect that most of the memory consumption can be attributed to that. I'm glad that it does work in the end though.

On x86_64 Linux, Hadolint consumes at its peak about 20Mb memory when compiled from source and about 38Mb when using the prebuilt release binaries. I think this may change depending on the input Dockerfile though.

**Comment by hpedrorodrigues** [comment]
Hi @m-ildefons, thanks again for taking the time to look at this issue.

As you pointed out Docker will utilize QEMU to run amd64 containers on M1 hardware, and as stated in the documentation running Intel-based containers on Arm-based machines should be regarded as "best-effort" only. Probably, even when it works it will be slower and use more memory.

---

Does GHC's linker support ARM architecture? More precisely, would it be possible to generate hadolint binaries for ARM? I'm sorry but I'm not familiar with Haskell ecosystem.

If the answers are yes, we can try to generate cross-platform images using [docker buildx](https://docs.docker.com/buildx/working-with-buildx/) (they are already configured in [docker workflow](https://github.com/hadolint/hadolint/blob/ab14d32dc6a8a951afa2363accc7c4ef2bba7d3c/.github/workflows/docker.yml#L23)), since there are official ARM images for Debian and Alpine. I can work on this but probably I'll need some guidance to compile the project and test things.

WDYT?

**Comment by lorenzo** [comment]
It is supported by GHC, we just want someone with time to change the CI to do it. Previous attempts has not been successful as they exceed the action time. We need separate actions.
